src/include/txn.i

```diff
@@ -186,11 +186,14 @@ __wt_txn_visible(WT_SESSION_IMPL *session, uint64_t id)
 		return (false);
 
 	/*
-	 * Read-uncommitted transactions see all other changes.
+	 * If we don't have a snapshot or we are running at read-uncommitted
+	 * isolation, all other changes are visible.
 	 */
 	if (txn->isolation == WT_ISO_READ_UNCOMMITTED)
 		return (true);
 
+	WT_ASSERT(session, F_ISSET(txn, WT_TXN_HAS_SNAPSHOT));
+
```

Comment:
Any visibility check that  isn't read-uncommitted should have an active snapshot.  This assertion found many paths that could give unexpected results (either when checkpointing the metadata or when evicting dirty files, which used to use the special `WT_ISO_EVICTION` isolation level.